articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,235 @@
+
+Service Servers and Clients will each specify two liveliness policies, one for the liveliness policy pertaining to the Server and one pertaining to the Client.
+In order for a Client to connect to a Server to make a request the Client_Liveliness level requested by the Server must be greater than the level provided by the Client and the Server_Liveliness requested by the Client must be greater than the level provided by the Server.
+
+### Lifespan
+
+The lifespan policy establishes a contract for how long a message remains valid.
+The lifespan QoS policy only applies to Topics.
+For Subscriptions it establishes the length of time a message is considered valid, after which time it will not be received.
+For Publishers it establishes the length of time a message is considered valid, after which time it will be removed from the Topic history and no longer sent to Subscribers.
+
+- LIFESPAN_DEFAULT - Use the ROS specified default for lifespan (which is LIFESPAN_NONE).
+- LIFESPAN_NONE - Messages do not have a time at which they expire.
+- LIFESPAN_ENABLED - Messages will have a lifespan enforced.
+
+In the future lifespan support could be added for Services to act as a means of timing out requests if they take over a set amount of time.
+At present, that would require additional support for Service Clients to be able to handle timeouts and cleanup requests that is out of scope for this design.
+
+### DDS QoS Relation
+
+These new policies are all based on the DDS QoS policies, but they do not require a DDS in order for an rmw implementation to support them.
+More detailed information on the DDS specifics of these policies can be found below in Appendix A.
+
+The only new QoS setting that does not directly map to DDS is the deadline policy for Services.
+While the deadline policy could map directly to the underlying Publisher and Subscriber like they do for Topics, that would tie the QoS policy to the DDS implementation instead of the generic Service definition that does not specify it be implemented using two Topics.
+The definition as it applies to messages on two underlying topics is also less useful than the definition of deadline as it pertains to life of a request.
+
+### ROS Changes
+
+These are the various changes that would be needed within ROS in order to natively support Deadline and Liveliness.
+
+#### Resource Status Event Handler
+
+Both the Deadline and Liveliness policies generate events from the rmw layer of which the application will need to be informed.
+For Deadlines, the Subscriber receives event notifications if it doesn't receive anything within the deadline and the Publisher receives event notifications if it doesn't publish anything within the deadline.
+For Liveliness, Subscribers receive events when the Publisher they're listening to is no longer alive.
```

Comment:
What if there are multiple publishers but each providing a subset of the necessary data model?

In DDS liveliness information is provided with the [SampleInfo](https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/api/connext_dds/api_cpp2/classdds_1_1sub_1_1SampleInfo.html)

Also, lifecycle impacts data buffering:
https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/api/connext_dds/api_cpp2/classdds_1_1core_1_1policy_1_1WriterDataLifecycle.html#details

Reply:
This should also be covered in a follow up issue (the same one as for deadline or in a separate one).

> Also, lifecycle impacts data buffering

I not sure what @deeplearningrobotics is speaking about here.

Reply:
@nburek I'm not sure what you mean by composition, since you can definitely have multiple publishers per subscription (many to one/many).

I think @deeplearningrobotics means that liveliness is based on the instance, so in the case that we have one instance (most of the time, i.e. with key-less types) the liveliness isn't lost unless all the publishers stop asserting liveliness. If at least one is still asserting it, then it will not drop. So it may not be a reliable way to detect that an individual publisher is gone, but instead that all publishers for a particular instance is gone.

Reply:
I think this may just need rewording here, because we're not doing anything special or different from DDS, we're just using a single instance. But that does not indicate a single publisher, which is kind of what this verbiage seems to be saying to me.

Reply:
I've made a change to the wording to indicate that the subscriber receives an event when there are no publishers alive to continue asserting liveliness on the topic. I did not specifically mention instances since it's assumed through the doc that ROS 2 only supports single instances and that multiple instances of a topic based on a key need further work.